include/boost/beast/websocket/detail/frame.hpp

```diff
@@ -89,159 +87,44 @@ is_control(opcode op)
 }
 
-inline
+BOOST_BEAST_DECL
 bool
-is_valid_close_code(std::uint16_t v)
-{
-    switch(v)
-    {
-    case close_code::normal:            // 1000
-    case close_code::going_away:        // 1001
-    case close_code::protocol_error:    // 1002
-    case close_code::unknown_data:      // 1003
-    case close_code::bad_payload:       // 1007
-    case close_code::policy_error:      // 1008
-    case close_code::too_big:           // 1009
-    case close_code::needs_extension:   // 1010
-    case close_code::internal_error:    // 1011
-    case close_code::service_restart:   // 1012
-    case close_code::try_again_later:   // 1013
-        return true;
-
-    // explicitly reserved
-    case close_code::reserved1:         // 1004
-    case close_code::no_status:         // 1005
-    case close_code::abnormal:          // 1006
-    case close_code::reserved2:         // 1014
-    case close_code::reserved3:         // 1015
-        return false;
-    }
-    // reserved
-    if(v >= 1016 && v <= 2999)
-        return false;
-    // not used
-    if(v <= 999)
-        return false;
-    return true;
-}
+is_valid_close_code(std::uint16_t v);
 
 //------------------------------------------------------------------------------
 
 // Write frame header to dynamic buffer
 //
-template<class DynamicBuffer>
+BOOST_BEAST_DECL
 void
-write(DynamicBuffer& db, frame_header const& fh)
-{
-    std::size_t n;
-    std::uint8_t b[14];
-    b[0] = (fh.fin ? 0x80 : 0x00) | static_cast<std::uint8_t>(fh.op);
-    if(fh.rsv1)
-        b[0] |= 0x40;
-    if(fh.rsv2)
-        b[0] |= 0x20;
-    if(fh.rsv3)
-        b[0] |= 0x10;
-    b[1] = fh.mask ? 0x80 : 0x00;
-    if(fh.len <= 125)
-    {
-        b[1] |= fh.len;
-        n = 2;
-    }
-    else if(fh.len <= 65535)
-    {
-        b[1] |= 126;
-        auto len_be = endian::native_to_big(
-            static_cast<std::uint16_t>(fh.len));
-        std::memcpy(&b[2], &len_be, sizeof(len_be));
-        n = 4;
-    }
-    else
-    {
-        b[1] |= 127;
-        auto len_be = endian::native_to_big(
-            static_cast<std::uint64_t>(fh.len));
-        std::memcpy(&b[2], &len_be, sizeof(len_be));
-        n = 10;
-    }
-    if(fh.mask)
-    {
-        auto key_le = endian::native_to_little(
-            static_cast<std::uint32_t>(fh.key));
-        std::memcpy(&b[n], &key_le, sizeof(key_le));
-        n += 4;
-    }
-    db.commit(net::buffer_copy(
-        db.prepare(n), net::buffer(b)));
-}
+write(flat_static_buffer_base& db, frame_header const& fh);
 
 // Read data from buffers
 // This is for ping and pong payloads
 //
-template<class Buffers>
+BOOST_BEAST_DECL
 void
-read_ping(ping_data& data, Buffers const& bs)
-{
-    BOOST_ASSERT(buffer_bytes(bs) <= data.max_size());
-    data.resize(buffer_bytes(bs));
-    net::buffer_copy(net::mutable_buffer{
-        data.data(), data.size()}, bs);
-}
+read_ping(
+    ping_data& data,
+    beast::buffers_prefix_view<beast::detail::buffers_pair<true>> const& bs);
```

Comment:
Hmmm.... this is dodgy. We shouldn't be leaking the type used by the implementation of the circular dynamic buffer (`buffers_pair`).

Reply:
Why not? We're already in `detail` and this is not observable to the user.